pkg/distributor/distributor.go

```diff
@@ -488,57 +560,74 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
 	const maxExpectedReplicationSet = 5 // typical replication factor 3 plus one for inactive plus one for luck
 	var descs [maxExpectedReplicationSet]ring.InstanceDesc
 
-	streamTrackers := make([]streamTracker, len(streams))
-	streamsByIngester := map[string][]*streamTracker{}
-	ingesterDescs := map[string]ring.InstanceDesc{}
+	tracker := pushTracker{
+		done: make(chan struct{}, 1), // buffer avoids blocking if caller terminates - sendSamples() only sends once on each
```

Comment:
I think it should have 2 buffer here if both are activate. Because it can lock down the sender side in case the context is cancelled.

Reply:
Same for error.

Reply:
In fact I don't think you can use a single pushTracker. Otherwise if one fails and one succeed, the
```
select {
	case err := <-tracker.err:
		return nil, err
	case <-tracker.done:
		return &logproto.PushResponse{}, validationErr
	case <-ctx.Done():
		return nil, ctx.Err()
	}
``` 

Will succeed.

I would use two of them and evaluate in a for loop both of them.

Reply:
I was wrong !

Reply:
discussed offline: I pulled the pushTracker logic into a single place to make it more obvious how it works. Only the first err or final push will write to the channels, so theres no danger of deadlock here.